### ISSUE TITLE
Undocumented behavior when following redirected PUT requests

### ISSUE THREAD

**Issue by mightyCelu** [curl/curl#11520]
### I did this

Issued a PUT request to a server that redirects via a 303 response:
```
$ curl -iLX 'PUT' --data 'content' localhost:1001/test
HTTP/1.1 303 See Other
Location: http://127.0.0.1:1002/test
Content-Length: 0

HTTP/1.1 201 Created
Content-Length: 0
```

### I expected the following

I expected `curl` to repeat the same `PUT` request to the server indicated by the return `Location` header.

Instead, the request is repeated, but without its payload. This does not only result in the intended operation/request not being issued, but instead leads to a different outcome than intended, as the resource is created, albeit without any content.

The behavior is similar to the documented behavior for POST requests that are redirected by 30{1,2,3} replies:

              When curl follows a redirect and if the request is a POST, it will send  the
              following  request  with a GET if the HTTP response was 301, 302, or 303. If
              the response code was any other 3xx code, curl will  re-send  the  following
              request using the same unmodified method.

              You can tell curl to not change POST requests to GET after a 30x response by
              using the dedicated options for that: --post301, --post302 and --post303.

In fact, passing `--post303` resolves this issue, So I assume it would also occur on 301/302 responses, and be resolved by the corresponding flags.
However, in contrast to the documented behavior on POST, though, instead of a GET request, a PUT request without payload is issued.
This leads to the aforementioned problem, as in contrast to GET, PUT is not a safe, i.e., side effect free, method.

### curl/libcurl version

curl 8.2.0 (x86_64-pc-linux-gnu) libcurl/8.2.0 OpenSSL/3.1.1 zlib/1.2.13 brotli/1.0.9 zstd/1.5.5 libidn2/2.3.4 libpsl/0.21.2 (+libidn2/2.3.4) libssh2/1.11.0 nghttp2/1.55.1 Release-Date: 2023-07-19 Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd

### operating system

Linux host 6.1.38-2-lts #1 SMP PREEMPT_DYNAMIC Sat, 15 Jul 2023 19:51:05 +0000 x86_64 GNU/Linux

**Comment by bagder** [comment]
I think parts of this is explained by the use of `-X PUT`. The `-X` option simply changes the method string in the request, it does not make curl act as a PUT. It still acts like `GET`, as that's the default. This is explained in the man page for `-X`.

`-T` is what you use to make curl _know_ it uses PUT.

**Comment by bagder** [comment]
I don't consider this a bug. This is `-X` doing its work.

**Comment by mightyCelu** [comment]
You're right,  I was not aware of the specifics of the `-X` flag, I should've read that better.
I still feel that it's too easy to easy to repeat my mistake, and that the consequences are potentially more severe than they should be.

**Comment by jzakrzewski** [comment]
Well, we discourage using `-X` whenever we can, but people either ignore this advise, or just copy-paste from places that ignored it. On top of that some tools that generate curl calls also use `-X` for no reason at all.

**Comment by MichalPetryka** [comment]
> Well, we discourage using `-X` whenever we can, but people either ignore this advise, or just copy-paste from places that ignored it.

Should curl maybe print a warning to stderr on such usage?

**Comment by dfandrich** [comment]
> Should curl maybe print a warning to stderr on such usage?

It does, but only in verbose mode.
